keps/sig-node/2400-node-swap/README.md

```diff
@@ -166,6 +172,83 @@ administrators can configure the kubelet such that:
+
+By doing so, we can ensure a thorough understanding of the feature's performance and stability before considering the manual input of swap values in a subsequent beta release. This cautious approach will ensure the efficient allocation of resources and the smooth integration of swap support into Kubernetes.
+
+Allocate the swap limit equal to the requested memory for each container and adjust the proportion of swap based on the total swap memory available.
+
+#### Set Aside Swap for System Critical Daemon
+
+System critical daemons (such as Kubelet) are essential for node health. Usually, an appropriate portion of system resources (e.g., memory, CPU) is reserved as system reserved. However, swap doesn't inherently support reserving a portion out of the total available. For instance, in the case of memory, we set `memory.min` on the node-level cgroup to ensure an adequate amount of memory is set aside, away from the pods, and for system critical daemons. But there is no equivalent for swap; i.e., no `memory.swap.min` is supported in the kernel. 
+
+Since this proposal advocates enabling swap only for the Burstable QoS pods, this can be done by setting `memory.swap.max` on the cgroups used by the Burstable QoS pods. The value of this `memory.swap.max` can be calculated by:
+
+memory.swap.max = total swap memory available on the system - system reserve (memory)
+
+This is the total amount of swap available for all the Burstable QoS pods; let's call it `TotalPodsSwapAvailable`. This will ensure that the system critical daemons will have access to the swap at least equal to the system reserved memory. This will indirectly act as having support for swap in system reserved.
+
+#### Use Swap Throttling
+
+`memory.swap.high` in the cgroup v2 allows you to enable throttling for swap allocation. If a cgroup’s swap usage exceeds `memory.swap.high`, all its further allocations will be throttled.
+
+Throttling will be enabled at individual Burstable QoS pod as well as at Burstable QoS cgroup level.
+
+1. At the Burstable QoS cgroup level, it can be calculated with: memory.swap.high = `TotalPodsSwapAvailable` * swap throttling factor
+2. At individual Burstable QoS pod level, it can be calculated with: memory.swap.high = memory.swap.max * swap throttling factor
+
+The default value of the memory throttling factor is set to 0.9.
+### Steps to Calculate Swap Limit
+
+1. **Calculate the total memory requests of the pod:**
+   - Sum up the memory requests of all containers in the pod. Let's call this value `TotalMemory`.
+
+2. **Determine the swap proportion for each container:**
```

Comment:
this may not be ideal that containers that do not request any memory (presuming those containers are not consuming much and happy to be swapped) will receive zero proportion of swap. I'd say those need to get some default or pre-defined proportion.

Reply:
Yes, this makes total sense.

The following approach comes into mind:
We can decide to allocate some percentage of the proportion, say 10% (we can discuss the exact value), that would split between **all** of the containers that do not request any memory.

<ins>For example:</ins>
For simplicity, let's assume that `SwapMemoryProportion` equals to 1 (i.e. swap size is equal to total physical memory size).

Let's say that a pod is defined with 4 containers:
* Container A: requests `1500Mi`
* Container B: requests `500Mi`
* Containers C, D and E: don't request memory.

The total pod memory: `2000Mi`. Then the `RequestedMemoryProportion` for the containers would be:
* Container A: `0.9 * (1500Mi / 2000Mi)` == `0.9 * 0.75` == `0.675`
* Container B: `0.9 * (500Mi / 2000Mi)` == `0.9 * 0.25` == `0.225`
* Container C and D: `0.1 * (1/3)` == `1/30` ~ `0.0333`

_note that `0.675+0.225+(3*0.0333) == 1`_

So, the total swap allocation for these containers would be:
* Container A: `0.675 * 2000Mi` == `1350Mi`
* Container B: `0.225 * 2000Mi` == `450Mi`
* Container C, D and E: `1/30 * 2000Mi` == `66.6667Mi`

_note that `1350Mi+450Mi+(3*66.6667Mi) == 2000Mi`_

WDYT?
FYI @harche

Reply:
@SergeyKanzhelev I am not sure if I got that correctly, but if a container needs swap then that container should request memory, that's the basic premise of this KEP. IMHO, this keeps things very simple and predictable.

Reply:
@SergeyKanzhelev I see, in that case we can probably go with @iholder101's [suggestion above](https://github.com/kubernetes/enhancements/pull/3957#discussion_r1219449643) or if you have a different idea to handle this, would like to see that as well. But I still believe, we are complicating things without gaining much. If we stay with the assumption that container gets swap only if requested, it is far simple approach.

Reply:
> Logically it makes sense. Practically I'm coming from the assumption that no request indicates "work on any leftover memory the Pod has". Meaning that the container is fine with whatever you can give to it. Which is a good candidate for swap memory.

IMHO having no request can also indicate "I'm not sure how much memory this container would consume", or even "this is a low priority pod, if you don't have enough memory for it, just kill it - it's fine by me".

Essentially, we're taking some of the swap allocation from containers that had explicitly requested it (by having memory requests) to containers that didn't, and I'm not sure that's the desired result for every use-case. Additionally, there is no appropriate way to disable swap for these containers, as setting request == limit might compromise the container of being killed even quicker, and is the opposite of the "best-effort" approach that was defined for it in the first place.

Lastly, I agree with @harche that it would complicate stuff. As a user, the calculation as it is now would seem much simpler to understand and manage, and one of the goals of this KEP is to keep configuration as transparent and simple as we can. I also don't see much of a disadvantage in defining some memory requests for these containers if I think they need swap to be allocated for them.

As @harche said, I'm fine with my approach from the previous comment, but I don't think it's necessary.

Let me know what you think.